AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -1176,23 +1176,28 @@ - (void)setContentsScaleForDisplay:(CGFloat)contentsScaleForDisplay
 - (void)applyPendingViewState
 {
   ASDisplayNodeAssertMainThread();
-  ASDN::MutexLocker l(__instanceLock__);
+  BOOL shouldSetNeedsLayout = NO;
+  {
+    ASDN::MutexLocker l(__instanceLock__);
+    shouldSetNeedsLayout = _pendingViewState.hasSetNeedsLayout;
+
+    if (self.layerBacked) {
+      [_pendingViewState applyToLayer:self.layer];
+    } else {
+      BOOL specialPropertiesHandling = ASDisplayNodeNeedsSpecialPropertiesHandlingForFlags(_flags);
+      [_pendingViewState applyToView:self.view withSpecialPropertiesHandling:specialPropertiesHandling];
+    }
 
+    [_pendingViewState clearChanges];
+  }
+
+  // NOTE: It's important to call this without holding the lock.
   // FIXME: Ideally we'd call this as soon as the node receives -setNeedsLayout
   // but automatic subnode management would require us to modify the node tree
   // in the background on a loaded node, which isn't currently supported.
-  if (_pendingViewState.hasSetNeedsLayout) {
+  if (shouldSetNeedsLayout) {
```

Comment:
Is it ok that setNeedsLayout has moved to after the pending state has been applied? Currently setNeedLayout sometimes lays out synchronously?